From 1131d880f3917c697abd64f1c79ddd8a9611286b Mon Sep 17 00:00:00 2001
From: Phil Elwell <phil@raspberrypi.com>
Date: Tue, 5 Jul 2022 17:00:50 +0100
Subject: [PATCH] overlays: Support i2c3-6 in container overlays

The I2C container overlays (i2c-rtc, i2c-fan and now i2c-sensor)
take parameters to select which I2C bus to use. Extend the list of
busses to include I2C3 to I2C6.

Note that this has been possible in a way that does prevent the overlay
from loading on older Pis due to the use of target-path = "i2c3"
instead of target = <&i2c3>. The latter fails on a non-BCM2711 Pi when
the overlay is loaded, whereas the former only fails when the fragment
in question is applied (and not at all if the fragment is dormant, or
path is created by a parameter that is not used).

Signed-off-by: Phil Elwell <phil@raspberrypi.com>
---
 arch/arm/boot/dts/overlays/README             |  40 ++
 .../arm/boot/dts/overlays/i2c-fan-overlay.dts |   8 +
 .../arm/boot/dts/overlays/i2c-rtc-overlay.dts |   8 +
 .../boot/dts/overlays/i2c-sensor-common.dtsi  | 341 ++++++++++++++++++
 .../boot/dts/overlays/i2c-sensor-overlay.dts  | 336 +----------------
 5 files changed, 416 insertions(+), 317 deletions(-)
 create mode 100755 arch/arm/boot/dts/overlays/i2c-sensor-common.dtsi

diff --git a/arch/arm/boot/dts/overlays/README b/arch/arm/boot/dts/overlays/README
index d85e8aabb7b3..a7300ac76ce9 100644
--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -1567,6 +1567,18 @@ Params: addr                    Sets the address for the fan controller. Note
 
         i2c_csi_dsi             Choose the I2C0 bus on GPIOs 44&45
 
+        i2c3                    Choose the I2C3 bus (configure with the i2c3
+                                overlay)
+
+        i2c4                    Choose the I2C3 bus (configure with the i2c3
+                                overlay)
+
+        i2c5                    Choose the I2C5 bus (configure with the i2c4
+                                overlay)
+
+        i2c6                    Choose the I2C6 bus (configure with the i2c6
+                                overlay)
+
         minpwm                  PWM setting for the fan when the SoC is below
                                 mintemp (range 0-255. default 0)
         maxpwm                  PWM setting for the fan when the SoC is above
@@ -1679,6 +1691,18 @@ Params: abx80x                  Select one of the ABx80x family:
 
         i2c_csi_dsi             Choose the I2C0 bus on GPIOs 44&45
 
+        i2c3                    Choose the I2C3 bus (configure with the i2c3
+                                overlay)
+
+        i2c4                    Choose the I2C3 bus (configure with the i2c3
+                                overlay)
+
+        i2c5                    Choose the I2C5 bus (configure with the i2c4
+                                overlay)
+
+        i2c6                    Choose the I2C6 bus (configure with the i2c6
+                                overlay)
+
         addr                    Sets the address for the RTC. Note that the
                                 device must be configured to use the specified
                                 address.
@@ -1845,6 +1869,22 @@ Params: addr                    Set the address for the BH1750, BME280, BME680,
         veml6070                Select the Vishay VEML6070 ultraviolet light
                                 sensor
 
+        i2c0                    Choose the I2C0 bus on GPIOs 0&1
+
+        i2c_csi_dsi             Choose the I2C0 bus on GPIOs 44&45
+
+        i2c3                    Choose the I2C3 bus (configure with the i2c3
+                                overlay)
+
+        i2c4                    Choose the I2C3 bus (configure with the i2c3
+                                overlay)
+
+        i2c5                    Choose the I2C5 bus (configure with the i2c4
+                                overlay)
+
+        i2c6                    Choose the I2C6 bus (configure with the i2c6
+                                overlay)
+
 
 Name:   i2c0
 Info:   Change i2c0 pin usage. Not all pin combinations are usable on all
diff --git a/arch/arm/boot/dts/overlays/i2c-fan-overlay.dts b/arch/arm/boot/dts/overlays/i2c-fan-overlay.dts
index e9424c376d54..947444136d99 100644
--- a/arch/arm/boot/dts/overlays/i2c-fan-overlay.dts
+++ b/arch/arm/boot/dts/overlays/i2c-fan-overlay.dts
@@ -77,6 +77,14 @@ __overrides__ {
 		i2c0 =		<&frag100>,"target:0=",<&i2c0>;
 		i2c_csi_dsi =	<&frag100>,"target:0=",<&i2c_csi_dsi>,
 				<0>,"+101+102";
+		i2c3 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c3";
+		i2c4 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c4";
+		i2c5 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c5";
+		i2c6 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c6";
 		addr =		<&emc2301>,"reg:0";
 		minpwm =	<&emc2301>,"emc2305,pwm-min;0";
 		maxpwm =	<&emc2301>,"emc2305,pwm-max;0";
diff --git a/arch/arm/boot/dts/overlays/i2c-rtc-overlay.dts b/arch/arm/boot/dts/overlays/i2c-rtc-overlay.dts
index 1eae9e1a5c96..cd31eac7e333 100644
--- a/arch/arm/boot/dts/overlays/i2c-rtc-overlay.dts
+++ b/arch/arm/boot/dts/overlays/i2c-rtc-overlay.dts
@@ -30,5 +30,13 @@ __overrides__ {
 		i2c0 = <&frag100>, "target:0=",<&i2c0>;
 		i2c_csi_dsi = <&frag100>, "target:0=",<&i2c_csi_dsi>,
 			      <0>,"+101+102";
+		i2c3 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c3";
+		i2c4 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c4";
+		i2c5 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c5";
+		i2c6 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c6";
 	};
 };
diff --git a/arch/arm/boot/dts/overlays/i2c-sensor-common.dtsi b/arch/arm/boot/dts/overlays/i2c-sensor-common.dtsi
new file mode 100755
index 000000000000..2716898e24a4
--- /dev/null
+++ b/arch/arm/boot/dts/overlays/i2c-sensor-common.dtsi
@@ -0,0 +1,341 @@
+// Definitions for I2C based sensors using the Industrial IO or HWMON interface.
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "brcm,bcm2835";
+
+	fragment@0 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			bme280: bme280@76 {
+				compatible = "bosch,bme280";
+				reg = <0x76>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@1 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			bmp085: bmp085@77 {
+				compatible = "bosch,bmp085";
+				reg = <0x77>;
+				default-oversampling = <3>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@2 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			bmp180: bmp180@77 {
+				compatible = "bosch,bmp180";
+				reg = <0x77>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@3 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			bmp280: bmp280@76 {
+				compatible = "bosch,bmp280";
+				reg = <0x76>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@4 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			htu21: htu21@40 {
+				compatible = "htu21";
+				reg = <0x40>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@5 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			lm75: lm75@4f {
+				compatible = "lm75";
+				reg = <0x4f>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@6 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			si7020: si7020@40 {
+				compatible = "si7020";
+				reg = <0x40>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@7 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			tmp102: tmp102@48 {
+				compatible = "ti,tmp102";
+				reg = <0x48>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@8 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			hdc100x: hdc100x@40 {
+				compatible = "hdc100x";
+				reg = <0x40>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@9 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			tsl4531: tsl4531@29 {
+				compatible = "tsl4531";
+				reg = <0x29>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@10 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			veml6070: veml6070@38 {
+				compatible = "veml6070";
+				reg = <0x38>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@11 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			sht3x: sht3x@44 {
+				compatible = "sht3x";
+				reg = <0x44>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@12 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			ds1621: ds1621@48 {
+				compatible = "ds1621";
+				reg = <0x48>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@13 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			max17040: max17040@36 {
+				compatible = "maxim,max17040";
+				reg = <0x36>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@14 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			bme680: bme680@76 {
+				compatible = "bosch,bme680";
+				reg = <0x76>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@15 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			sps30: sps30@69 {
+				compatible = "sensirion,sps30";
+				reg = <0x69>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@16 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			sgp30: sgp30@58 {
+				compatible = "sensirion,sgp30";
+				reg = <0x58>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@17 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			ccs811: ccs811@5b {
+				compatible = "ccs811";
+				reg = <0x5b>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@18 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			bh1750: bh1750@23 {
+				compatible = "bh1750";
+				reg = <0x23>;
+				status = "okay";
+			};
+		};
+	};
+
+	fragment@19 {
+		target = <&i2cbus>;
+		__dormant__ {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			status = "okay";
+
+			max30102: max30102@57 {
+				compatible = "maxim,max30102";
+				reg = <0x57>;
+				maxim,red-led-current-microamp = <7000>;
+				maxim,ir-led-current-microamp  = <7000>;
+				interrupt-parent = <&gpio>;
+				interrupts = <4 2>;
+			};
+		};
+	};
+
+	__overrides__ {
+		bme280 = <0>,"+0";
+		bmp085 = <0>,"+1";
+		bmp180 = <0>,"+2";
+		bmp280 = <0>,"+3";
+		htu21 = <0>,"+4";
+		lm75 = <0>,"+5";
+		lm75addr = <&lm75>,"reg:0";
+		si7020 = <0>,"+6";
+		tmp102 = <0>,"+7";
+		hdc100x = <0>,"+8";
+		tsl4531 = <0>,"+9";
+		veml6070 = <0>,"+10";
+		sht3x = <0>,"+11";
+		ds1621 = <0>,"+12";
+		max17040 = <0>,"+13";
+		bme680 = <0>,"+14";
+		sps30 = <0>,"+15";
+		sgp30 = <0>,"+16";
+		ccs811 = <0>, "+17";
+		bh1750 = <0>, "+18";
+		max30102 = <0>,"+19";
+
+		addr =  <&bme280>,"reg:0", <&bmp280>,"reg:0", <&tmp102>,"reg:0",
+			<&lm75>,"reg:0", <&hdc100x>,"reg:0", <&sht3x>,"reg:0",
+			<&ds1621>,"reg:0", <&bme680>,"reg:0", <&ccs811>,"reg:0",
+			<&bh1750>,"reg:0";
+		int_pin = <&max30102>, "interrupts:0";
+	};
+};
diff --git a/arch/arm/boot/dts/overlays/i2c-sensor-overlay.dts b/arch/arm/boot/dts/overlays/i2c-sensor-overlay.dts
index 33965be4b1e8..f8a39659d83e 100755
--- a/arch/arm/boot/dts/overlays/i2c-sensor-overlay.dts
+++ b/arch/arm/boot/dts/overlays/i2c-sensor-overlay.dts
@@ -2,339 +2,41 @@
 /dts-v1/;
 /plugin/;
 
-/ {
-	compatible = "brcm,bcm2835";
-
-	fragment@0 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			bme280: bme280@76 {
-				compatible = "bosch,bme280";
-				reg = <0x76>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@1 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			bmp085: bmp085@77 {
-				compatible = "bosch,bmp085";
-				reg = <0x77>;
-				default-oversampling = <3>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@2 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			bmp180: bmp180@77 {
-				compatible = "bosch,bmp180";
-				reg = <0x77>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@3 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			bmp280: bmp280@76 {
-				compatible = "bosch,bmp280";
-				reg = <0x76>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@4 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			htu21: htu21@40 {
-				compatible = "htu21";
-				reg = <0x40>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@5 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			lm75: lm75@4f {
-				compatible = "lm75";
-				reg = <0x4f>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@6 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			si7020: si7020@40 {
-				compatible = "si7020";
-				reg = <0x40>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@7 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			tmp102: tmp102@48 {
-				compatible = "ti,tmp102";
-				reg = <0x48>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@8 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			hdc100x: hdc100x@40 {
-				compatible = "hdc100x";
-				reg = <0x40>;
-				status = "okay";
-			};
-		};
-	};
+#include "i2c-sensor-common.dtsi"
 
-	fragment@9 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			tsl4531: tsl4531@29 {
-				compatible = "tsl4531";
-				reg = <0x29>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@10 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			veml6070: veml6070@38 {
-				compatible = "veml6070";
-				reg = <0x38>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@11 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			sht3x: sht3x@44 {
-				compatible = "sht3x";
-				reg = <0x44>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@12 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			ds1621: ds1621@48 {
-				compatible = "ds1621";
-				reg = <0x48>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@13 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			max17040: max17040@36 {
-				compatible = "maxim,max17040";
-				reg = <0x36>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@14 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			bme680: bme680@76 {
-				compatible = "bosch,bme680";
-				reg = <0x76>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@15 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			sps30: sps30@69 {
-				compatible = "sensirion,sps30";
-				reg = <0x69>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@16 {
-		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
-			status = "okay";
-
-			sgp30: sgp30@58 {
-				compatible = "sensirion,sgp30";
-				reg = <0x58>;
-				status = "okay";
-			};
-		};
-	};
-
-	fragment@17 {
+/ {
+	frag100: fragment@100 {
 		target = <&i2c_arm>;
-		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
+		i2cbus: __overlay__ {
 			status = "okay";
-
-			ccs811: ccs811@5b {
-				compatible = "ccs811";
-				reg = <0x5b>;
-				status = "okay";
-			};
 		};
 	};
 
-	fragment@18 {
-		target = <&i2c_arm>;
+	fragment@101 {
+		target = <&i2c0if>;
 		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
 			status = "okay";
-
-			bh1750: bh1750@23 {
-				compatible = "bh1750";
-				reg = <0x23>;
-				status = "okay";
-			};
 		};
 	};
 
-	fragment@19 {
-		target = <&i2c_arm>;
+	fragment@102 {
+		target = <&i2c0mux>;
 		__dormant__ {
-			#address-cells = <1>;
-			#size-cells = <0>;
 			status = "okay";
-
-			max30102: max30102@57 {
-				compatible = "maxim,max30102";
-				reg = <0x57>;
-				maxim,red-led-current-microamp = <7000>;
-				maxim,ir-led-current-microamp  = <7000>;
-				interrupt-parent = <&gpio>;
-				interrupts = <4 2>;
-			};
 		};
 	};
 
 	__overrides__ {
-		addr =  <&bme280>,"reg:0", <&bmp280>,"reg:0", <&tmp102>,"reg:0",
-			<&lm75>,"reg:0", <&hdc100x>,"reg:0", <&sht3x>,"reg:0",
-			<&ds1621>,"reg:0", <&bme680>,"reg:0", <&ccs811>,"reg:0",
-			<&bh1750>,"reg:0";
-		int_pin = <&max30102>, "interrupts:0";
-		bme280 = <0>,"+0";
-		bmp085 = <0>,"+1";
-		bmp180 = <0>,"+2";
-		bmp280 = <0>,"+3";
-		htu21 = <0>,"+4";
-		lm75 = <0>,"+5";
-		lm75addr = <&lm75>,"reg:0";
-		si7020 = <0>,"+6";
-		tmp102 = <0>,"+7";
-		hdc100x = <0>,"+8";
-		tsl4531 = <0>,"+9";
-		veml6070 = <0>,"+10";
-		sht3x = <0>,"+11";
-		ds1621 = <0>,"+12";
-		max17040 = <0>,"+13";
-		bme680 = <0>,"+14";
-		sps30 = <0>,"+15";
-		sgp30 = <0>,"+16";
-		ccs811 = <0>, "+17";
-		bh1750 = <0>, "+18";
-		max30102 = <0>,"+19";
+		i2c0 = <&frag100>, "target:0=",<&i2c0>;
+		i2c_csi_dsi = <&frag100>, "target:0=",<&i2c_csi_dsi>,
+			      <0>,"+101+102";
+		i2c3 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c3";
+		i2c4 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c4";
+		i2c5 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c5";
+		i2c6 = <&frag100>, "target?=0",
+		       <&frag100>, "target-path=i2c6";
 	};
 };
-- 
2.32.0

